File: DP/#5 - Longest Palindromic Substring - Medium/Solution.py
# Approach 1: Brute Force
class SolutionBruteForce:
    def longestPalindrome(self, s: str) -> str:
        def is_palindrome(sub):
            return sub == sub[::-1]
        res = ""
        n = len(s)
        for i in range(n):
            for j in range(i, n):
                if is_palindrome(s[i:j+1]) and len(s[i:j+1]) > len(res):
                    res = s[i:j+1]
        return res


# Approach 2: Expand Around Center
class SolutionExpand:
    def longestPalindrome(self, s: str) -> str:
        def expand(l, r):
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            return s[l+1:r]
        
        res = ""
        for i in range(len(s)):
            odd = expand(i, i)
            if len(odd) > len(res): res = odd
            even = expand(i, i+1)
            if len(even) > len(res): res = even
        return res


# Approach 3: Dynamic Programming
class SolutionDP:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        res = ""
        
        for i in range(n-1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i < 3 or dp[i+1][j-1]):
                    dp[i][j] = True
                if dp[i][j] and (j - i + 1) > len(res):
                    res = s[i:j+1]
        return res

File: DP/#5 - Longest Palindromic Substring - Medium/Solution.java
// Approach 1: Brute Force
class SolutionBruteForce {
    private boolean isPalindrome(String str) {
        int l = 0, r = str.length() - 1;
        while (l < r) {
            if (str.charAt(l++) != str.charAt(r--)) return false;
        }
        return true;
    }

    public String longestPalindrome(String s) {
        String res = "";
        int n = s.length();
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                String sub = s.substring(i, j + 1);
                if (isPalindrome(sub) && sub.length() > res.length()) {
                    res = sub;
                }
            }
        }
        return res;
    }
}

// Approach 2: Expand Around Center
class SolutionExpand {
    private String expand(String s, int l, int r) {
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l--; r++;
        }
        return s.substring(l+1, r);
    }

    public String longestPalindrome(String s) {
        String res = "";
        for (int i = 0; i < s.length(); i++) {
            String odd = expand(s, i, i);
            if (odd.length() > res.length()) res = odd;
            String even = expand(s, i, i+1);
            if (even.length() > res.length()) res = even;
        }
        return res;
    }
}

// Approach 3: Dynamic Programming
class SolutionDP {
    public String longestPalindrome(String s) {
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        String res = "";
        
        for (int i = n-1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s.charAt(i) == s.charAt(j) && (j - i < 3 || dp[i+1][j-1])) {
                    dp[i][j] = true;
                }
                if (dp[i][j] && (j - i + 1) > res.length()) {
                    res = s.substring(i, j+1);
                }
            }
        }
        return res;
    }
}
